core/vm/contracts.go

```diff
@@ -642,17 +661,30 @@ func (c *numberValidators) RequiredGas(input []byte) uint64 {
 // Return the number of validators that are required to sign this current, possibly unsealed, block. If this block is
 // the last in an epoch, note that that may mean one or more of those validators may no longer be elected
 // for subsequent blocks.
+// WARNING: Validator set is always constructed from the canonical chain, therefore this precompile is undefined
+// if the engine is aware of a chain with higher total difficulty.
 func (c *numberValidators) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
 	gas, err := debitRequiredGas(c, input, gas)
 	if err != nil {
 		return nil, gas, err
 	}
 
-	if len(input) != 0 {
+	// input is comprised of a single argument:
+	//   blockNumber: 32 byte integer representing the block number to access
+	if len(input) < 32 {
 		return nil, gas, ErrInputLength
 	}
 
-	validators := evm.Context.Engine.GetValidators(big.NewInt(evm.Context.BlockNumber.Int64()-1), evm.Context.GetHash(evm.Context.BlockNumber.Uint64()-1))
+	blockNumber := new(big.Int).SetBytes(input[0:32])
+	if blockNumber.Cmp(common.Big0) == 0 {
+		// Genesis validator set is empty. Return 0.
```

Comment:
It returns an error in the other precompile because when the validator set is empty, as in the case of the genesis block, every index is out of bounds. The other precompile was returning the wrong error, so I fixed that and added a comment.